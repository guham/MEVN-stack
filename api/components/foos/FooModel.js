const mongoose = require('mongoose');

const schema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
      unique: true,
    },
  },
  {
    timestamps: true,
  },
);

module.exports = mongoose.model('Foo', schema);
