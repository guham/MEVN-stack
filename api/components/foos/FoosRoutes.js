const { Router } = require('express');
const { makeInvoker } = require('awilix-express');
const { asyncMiddleware } = require('../../middlewares/errorHandlers');
const foosController = require('./FoosController');

const router = Router();
const api = makeInvoker(foosController);

router.get('/', asyncMiddleware(api('index')));
router.post('/add', asyncMiddleware(api('add')));
router.get('/test', asyncMiddleware(api('test')));
router.get('/:id', asyncMiddleware(api('get')));

module.exports = router;
