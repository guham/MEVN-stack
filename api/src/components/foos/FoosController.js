const foosController = ({ foosService }) => ({
  index: async (req, res, next) => {
    const foos = await foosService.getAll();
    res.json(foos);
  },

  add: async (req, res, next) => {
    const { name } = req.body;
    const foo = await foosService.add([
      { name },
    ]);
    res.status(201).json(foo);
  },

  get: async (req, res, next) => {
    const { id } = req.params;
    const foo = await foosService.getById(id);
    res.json(foo);
  },

  test: async (req, res, next) => {
    res.json({ data: 'This is a message fetched from the API' });
  },
});

module.exports = foosController;
