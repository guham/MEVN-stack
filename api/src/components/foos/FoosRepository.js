class FoosRepository {
  constructor({ FooModel }) {
    this.FooModel = FooModel;
  }

  async find(conditions = {}) {
    return this.FooModel.find(conditions).exec();
  }

  async add(documents) {
    return this.FooModel.create(...documents);
  }

  async getById(id) {
    return this.FooModel.findById(id).exec();
  }
}

module.exports = FoosRepository;
