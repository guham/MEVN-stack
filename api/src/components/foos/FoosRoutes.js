const { Router } = require('express');
const { makeInvoker } = require('awilix-express');
const { asyncHandler } = require('../../middlewares');
const foosController = require('./FoosController');

const router = Router();
const api = makeInvoker(foosController);

router.get('/', asyncHandler(api('index')));
router.post('/add', asyncHandler(api('add')));
router.get('/test', asyncHandler(api('test')));
router.get('/:id', asyncHandler(api('get')));

module.exports = router;
