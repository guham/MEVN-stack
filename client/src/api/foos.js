import { createClient } from './client';

const client = createClient();

export default {
  /**
   * @returns {Array<Object>} array of foos
   */
  async fetchFoos() {
    const response = await client.get('/api/foo');
    return response.data;
  },
  /**
   * @param {string} name
   * @returns {Object} added foo
   */
  async addFoo(name) {
    const response = await client.post('/api/foo/add', { name });
    return response.data;
  },
};
