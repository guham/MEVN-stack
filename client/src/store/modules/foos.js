import _get from 'lodash/get';
import foosClient from '@/api/foos';
import * as types from '../mutation-types';

const state = {
  foos: [],
  name: '',
  error: {},
};

const getters = {
  count: state => state.foos.length,
  isValidName: state => state.name.length > 0,
  hasError: state => Object.keys(state.error).length > 0,
  errorMessage: (state, getters) => (getters.hasError ? _get(state.error, 'data.message', _get(state.error, 'message', '')) : ''),
};

const mutations = {
  [types.FETCH_FOOS](state, foos) {
    state.foos = foos;
  },

  [types.ADD_FOO](state, foo) {
    state.foos.push(foo);
  },

  [types.UPDATE_FOO_NAME](state, name) {
    state.name = name.trim();
  },

  [types.SET_ERROR](state, error) {
    state.error = error;
  },

  [types.RESET_ERROR](state) {
    state.error = {};
  },
};

const actions = {
  async fetchFoos({ commit }) {
    try {
      const foos = await foosClient.fetchFoos();
      commit(types.FETCH_FOOS, foos);
    } catch (error) {
      commit(types.SET_ERROR, error);
    }
  },

  async addFoo({ commit, state, getters }) {
    if (!getters.isValidName) {
      return;
    }
    try {
      const foo = await foosClient.addFoo(state.name);
      commit(types.ADD_FOO, foo);
      commit(types.UPDATE_FOO_NAME, '');
      commit(types.RESET_ERROR);
    } catch (error) {
      commit(types.SET_ERROR, error);
    }
  },
};

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
};
