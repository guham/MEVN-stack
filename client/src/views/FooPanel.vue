<template>
  <div>
    <h1 class="pb-2">FooPanel</h1>
    <p class="py-2">Count: {{ count }}</p>
    <div class="flex">
      <div class="sm:w-px lg:w-2/3"/>
      <add-foo placeholder="Foo name"/>
      <div class="sm:w-px lg:w-2/3"/>
    </div>
  </div>
</template>

<script>
import { createNamespacedHelpers } from 'vuex';
import AddFoo from '@/components/AddFoo.vue';

const { mapGetters, mapActions } = createNamespacedHelpers('foos');

export default {
  name: 'FooPanel',
  components: {
    AddFoo,
  },
  data() {
    return {};
  },

  computed: {
    ...mapGetters([
      'count',
    ]),
  },

  created() {
    this.fetchFoos();
  },

  methods: {
    ...mapActions([
      'fetchFoos',
    ]),
  },
};
</script>

