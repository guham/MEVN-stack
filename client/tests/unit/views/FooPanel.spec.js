import Vuex from 'vuex';
import axios from 'axios';
import FooPanel from '@/views/FooPanel.vue';
import foos from '@/store/modules/foos';
import factory from '../factory';

const store = new Vuex.Store({
  modules: {
    foos,
  },
});

beforeEach(() => {
  jest.resetModules();
  jest.clearAllMocks();
});

axios.get.mockImplementation(() => Promise.resolve({
  data: [
    {
      _id: '5afa40240b166302e57ca0a0',
      name: 'name 1',
      createdAt: '2018-05-14T02:04:20.561Z',
      updatedAt: '2018-05-15T06:12:25.561Z',
      __v: 0,
    },
    {
      _id: '5afa446c1870de01d68fe0d6',
      name: 'name 2',
      createdAt: '2018-05-15T03:35:51.523Z',
      updatedAt: '2018-05-15T03:35:51.523Z',
      __v: 0,
    },
  ],
}));

describe('FooPanel.vue', () => {
  test('when the view is initialized, there are no foos', () => {
    const wrapper = factory(FooPanel, store);
    expect(wrapper.find('p').text()).toBe('Count: 0');
  });

  test('has a created hook', () => {
    expect(typeof FooPanel.created).toBe('function');
  });

  test('once created, an API call is done', async () => {
    const wrapper = factory(FooPanel, store);
    await wrapper.vm.fetchFoos();
    expect(axios.get).toBeCalled();
  });

  test('API endpoint should be called with "/api/foo"', async () => {
    const wrapper = factory(FooPanel, store);
    await wrapper.vm.fetchFoos();
    expect(axios.get).toHaveBeenCalledWith('/api/foo', {});
  });

  test('once created, renders foos count value', async () => {
    const wrapper = factory(FooPanel, store);
    await wrapper.vm.fetchFoos();
    expect(wrapper.find('p').text()).toBe('Count: 2');
  });

  test('has the expected html structure', () => {
    const wrapper = factory(FooPanel, store);
    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
