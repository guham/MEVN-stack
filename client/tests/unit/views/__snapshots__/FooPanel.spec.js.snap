// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooPanel.vue has the expected html structure 1`] = `
<div>
  <h1 class="pb-2 text-center">
    FooPanel
  </h1>
  <p class="py-2 text-center">
    total count
  </p>
  <addfoo-stub placeholder="placeholder"></addfoo-stub>
</div>
`;
